rotkehlchen/chain/evm/decoding/gearbox/gearbox_cache.py

```diff
@@ -0,0 +1,305 @@
+    and the special ETH address."""
+    try:
+        return {
+            de_addr
+            for data in pool_data[23]  # ZapperInfo[]
+            for addr in data[1:]  # ignore zapper address, only consider tokenIn and tokenOut
+            if (de_addr := deserialize_evm_address(addr)) not in {underlying_token_address, ETH_SPECIAL_ADDRESS}  # noqa: E501
+        }
+    except (DeserializationError, IndexError) as e:
+        log.error(f'Could not deserialize zapper tokens from {inquirer.chain_name} gearbox pool data with underlying as {underlying_token_address}: {e}')  # noqa: E501
+        return None
+
+
+def query_gearbox_data_from_chain(
+        evm_inquirer: 'EvmNodeInquirer',
+        existing_pools: set[ChecksumEvmAddress],
+) -> list[GearboxPoolData] | None:
+    """
+    Query all Gearbox information(lp tokens, pools, lp coins) from data compressor.
+
+    May raise:
+    - RemoteError if failed to query chain
+    """
+    data_compressor = evm_inquirer.contracts.contract(GEARBOX_DATA_COMPRESSOR)
+    pools_data = data_compressor.call(
+        node_inquirer=evm_inquirer,
+        method_name='getPoolsV3List',
+    )
+    gearbox_farming_pool_abi = evm_inquirer.contracts.abi('GEARBOX_FARMING_POOL')
+    new_pools = []
+    for pool_data in pools_data:
+        try:
+            pool_address = deserialize_evm_address(pool_data[0])
+        except DeserializationError:
+            log.error(f'Could not deserialize evm address {pool_data[0]} while decoding gearbox pool information')  # noqa: E501
+            continue
+
+        if pool_address in existing_pools:
+            continue
+
+        farming_pool_ids: set[str] = set()
+        lp_token_ids: set[str] = set()
+        try:
+            pool_name: str = pool_data[4]
+            underlying_token_address = deserialize_evm_address(pool_data[1])
+        except (DeserializationError, IndexError) as e:
+            log.error(f'Could not deserialize evm address while decoding gearbox pool {pool_address} information from data_compressor: {e}')  # noqa: E501
+            continue
+
+        zapper_tokens = get_zapper_tokens(evm_inquirer, pool_data, underlying_token_address)
+        if zapper_tokens is None:
+            log.error(f'Could not get zapper tokens for gearbox pool {pool_address} on {evm_inquirer.chain_name}')  # noqa: E501
+            continue
+
+        calls = [
+            (
+                (_contract := EvmContract(
+                    address=address,
+                    abi=gearbox_farming_pool_abi,
+                    deployed_block=0,
+                )).address,
+                _contract.encode(method_name='stakingToken'),
+            )
+            for address in zapper_tokens
+        ]
+        output = evm_inquirer.multicall_2(require_success=False, calls=calls)
+        for token_address, (is_farming_pool, _), in zip(zapper_tokens, output, strict=True):
+            if is_farming_pool:
+                farming_pool_ids.add(token_address)
```

Comment:
do you really need this to be a set. It's an error if it's not 1. Shouldn't you just find and save it? You even check if its a set of length 1 below and then even go so far as popping from the set to get that single element. So I don't see why you need to use a set.